File: apps/hypercortex/src/notePackagePaths.ts
import { NOTES_DIR, noteMonthFolderFromIdOrNow } from './core'

export const TRASH_DIR = 'Trash'

export function normalizeVaultPath(path: string): string {
  return String(path || '').trim().replace(/\\/g, '/').replace(/\/+/g, '/').replace(/^\/+|\/+$/g, '')
}

export function notePackageDirForId(id: string): string {
  const noteId = normalizeNotePackageId(id)
  return `${NOTES_DIR}/${noteMonthFolderFromIdOrNow(noteId)}/${noteId}`
}

export function notePathInPackage(packageDir: string, file: string): string {
  const dir = pathParts(packageDir).join('/')
  const name = normalizePathSegment(file, '文件名')
  return `${dir}/${name}`
}

export function noteDirToTrashDir(noteDir: string): string {
  const parts = notePackagePathParts(noteDir, NOTES_DIR, '移入回收站')
  return [TRASH_DIR, ...parts.slice(1)].join('/')
}

export function trashDirToNoteDir(trashDir: string): string {
  const parts = notePackagePathParts(trashDir, TRASH_DIR, '恢复')
  return [NOTES_DIR, ...parts.slice(1)].join('/')
}

export function canonicalOriginalDirForTrashPackage(trashPackageDir: string, originalDir: string | undefined, noteId: string): string {
  const source = String(originalDir || '').trim() || trashDirToNoteDir(trashPackageDir)
  return canonicalNoteDirForId(source, noteId)
}

export function canonicalNoteDirForId(noteDir: string, noteId: string): string {
  const id = normalizeNotePackageId(noteId)
  const parts = notePackagePathParts(noteDir, NOTES_DIR, '规范化')
  return [NOTES_DIR, parts[1], id].join('/')
}

function notePackagePathParts(path: string, expectedRoot: string, action: string): string[] {
  const parts = pathParts(path)
  if (parts.length !== 3 || parts[0] !== expectedRoot) throw new Error(`笔记目录无法${action}：${normalizeVaultPath(path)}`)
  return parts
}

function normalizeNotePackageId(id: string): string {
  const noteId = String(id || '').trim()
  if (!noteId) throw new Error('笔记 id 不能为空')
  if (noteId === '.' || noteId === '..' || /[\\/\u0000]/.test(noteId)) throw new Error(`笔记 id 不能作为文件夹名：${noteId}`)
  return noteId
}

function normalizePathSegment(value: string, label: string): string {
  const segment = String(value || '').trim()
  if (!segment) throw new Error(`${label}不能为空`)
  if (segment === '.' || segment === '..' || /[\\/\u0000]/.test(segment)) throw new Error(`${label}不能包含路径分隔符`)
  return segment
}

function pathParts(path: string): string[] {
  const normalized = normalizeVaultPath(path)
  if (!normalized) throw new Error('路径不能为空')
  const parts = normalized.split('/')
  if (parts.some(part => !part || part === '.' || part === '..' || /\u0000/.test(part))) throw new Error(`路径包含非法片段：${normalized}`)
  return parts
}

File: apps/hypercortex/backend-go/note_package_paths.go
package main

import (
	"errors"
	"fmt"
	"path/filepath"
	"strings"
	"time"
)

func noteMonthFolderFromID(id string) string {
	s := strings.TrimSpace(id)
	if len(s) >= 6 {
		y, m := s[:4], s[4:6]
		if allDigits(y) && allDigits(m) && m >= "01" && m <= "12" {
			return y + "-" + m
		}
	}
	return time.Now().Format("2006-01")
}

func notePackageDirForID(id string) (string, error) {
	noteID, err := normalizeNotePackageID(id)
	if err != nil {
		return "", err
	}
	return filepath.ToSlash(filepath.Join(notesDir, noteMonthFolderFromID(noteID), noteID)), nil
}

func notePackageDirForMonthAndID(rootName string, monthName string, id string) (string, error) {
	root := strings.TrimSpace(rootName)
	month := strings.TrimSpace(monthName)
	if root != notesDir && root != trashDir {
		return "", fmt.Errorf("非法笔记包根目录：%s", root)
	}
	if month == "" || strings.ContainsAny(month, "/\\\x00") || month == "." || month == ".." {
		return "", fmt.Errorf("非法笔记月份目录：%s", month)
	}
	noteID, err := normalizeNotePackageID(id)
	if err != nil {
		return "", err
	}
	return filepath.ToSlash(filepath.Join(root, month, noteID)), nil
}

func canonicalOriginalDirForTrashPackage(trashPackageDir string, originalDir string, noteID string) (string, error) {
	source := strings.TrimSpace(originalDir)
	if source == "" {
		parts, err := notePackagePathParts(trashPackageDir, trashDir, "恢复")
		if err != nil {
			return "", err
		}
		source = filepath.ToSlash(filepath.Join(notesDir, parts[1], parts[2]))
	}
	return canonicalNoteDirForID(source, noteID)
}

func canonicalNoteDirForID(dir string, noteID string) (string, error) {
	noteID, err := normalizeNotePackageID(noteID)
	if err != nil {
		return "", err
	}
	parts, err := notePackagePathParts(dir, notesDir, "规范化")
	if err != nil {
		return "", err
	}
	return notePackageDirForMonthAndID(notesDir, parts[1], noteID)
}

func trashPackageDirForNoteDir(noteDir string) (string, error) {
	parts, err := notePackagePathParts(noteDir, notesDir, "移入回收站")
	if err != nil {
		return "", err
	}
	return filepath.ToSlash(filepath.Join(trashDir, parts[1], parts[2])), nil
}

func notePackagePathParts(path string, expectedRoot string, action string) ([]string, error) {
	clean, err := cleanRelPath(path)
	if err != nil {
		return nil, err
	}
	parts := strings.Split(filepath.ToSlash(clean), "/")
	if len(parts) != 3 || parts[0] != expectedRoot {
		return nil, fmt.Errorf("笔记目录无法%s：%s", action, filepath.ToSlash(clean))
	}
	return parts, nil
}

func normalizeNotePackageID(id string) (string, error) {
	noteID := strings.TrimSpace(id)
	if noteID == "" {
		return "", errors.New("笔记 id 不能为空")
	}
	if noteID == "." || noteID == ".." || strings.ContainsAny(noteID, "/\\\x00") {
		return "", fmt.Errorf("笔记 id 不能作为文件夹名：%s", noteID)
	}
	return noteID, nil
}

func allDigits(s string) bool {
	for _, r := range s {
		if r < '0' || r > '9' {
			return false
		}
	}
	return s != ""
}
